pkg/auto/auto.go

```diff
@@ -0,0 +1,237 @@
+	if sameCharts {
+		return false
+	}
+
+	// Check for special edge cases
+	edgeCase := checkEdgeCasesIfChartChanged(lastChart, currentChart)
+	if !edgeCase {
+		return false
+	}
+
+	return true
+}
+
+// checkEdgeCasesIfChartChanged will check for special cases like:
+//
+//	-CRD dependencies
+//	-fleet
+//	-neuvector
+//	-rancher-alerting-driver + rancher-aks-operator
+//	-rancher-gke-operator + rancher-gatekeeper
+//
+// It will return true if the chart changed, false otherwise
+func checkEdgeCasesIfChartChanged(lastChart, currentChart string) bool {
+
+	lastParts := strings.Split(lastChart, "-")
+	currentParts := strings.Split(currentChart, "-")
+
+	minLength := 0 // compare which chart is shorter in name (last or current)
+	if len(currentParts) < len(lastParts) {
+		minLength = len(currentParts)
+	} else {
+		minLength = len(lastParts)
+	}
+
+	equalCounter := 0
+	for i := 0; i < minLength; i++ {
+		if lastParts[i] == currentParts[i] {
+			equalCounter++
+		}
+	}
+
+	if equalCounter == 0 {
+		return true
+	}
+	if equalCounter == 1 && strings.HasPrefix("rancher-", lastParts[0]) &&
+		minLength < 3 {
+		return true
+	}
+
+	if equalCounter == 1 && minLength >= 3 {
+		return true
+	}
+
+	// treat operators edge cases
+	// rancher-aks-operator && rancher-gke-operator || rancher-aks-operator-crd && rancher-gke-operator-crd
+	if (equalCounter == 2 && minLength == 3) || (equalCounter == 3 && minLength == 4) {
+		if lastParts[0] == "rancher" && currentParts[0] == "rancher" {
+			if lastParts[2] == "operator" && currentParts[2] == "operator" {
+				if lastParts[1] != currentParts[1] {
+					return true
+				}
+			}
+		}
+	}
+
+	return false
+}
+
+// createNewBranchToForwardPort will create a new branch to forward-port the assets
+func (fp *ForwardPort) createNewBranchToForwardPort(branch string) error {
+	// check if git is clean and branch is up-to-date
+	err := fp.git.IsClean()
+	if err != nil {
+		return err
+	}
+	// create new branch and checkout
+	err = fp.git.CreateAndCheckoutBranch(branch)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// prepareReleaseYaml will prepare the release.yaml file by erasing its content,
+// this is a good practice before releasing or forward-porting any charts with it.
+func prepareReleaseYaml() error {
+	// Check if the file exists
+	_, err := os.Stat("release.yaml")
+	if os.IsNotExist(err) {
+		logrus.Error("release.yaml does not exist.")
+		return fmt.Errorf("release.yaml does not exist")
+	} else if err != nil {
+		return err // return any other error encountered
+	}
+
+	// File exists, open it with O_TRUNC to erase its content
+	file, err := os.OpenFile("release.yaml", os.O_WRONLY|os.O_TRUNC, 0644)
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+
+	logrus.Info("Content of release.yaml erased successfully.")
+	return nil
+}
+
+// executeCommand will execute the given command using the yqPath if needed
+func executeCommand(command []string, yqPath string) error {
```

Comment:
Why aren't we using any of the yaml parser libraries available?

Reply:
Agreed, I will do some research on this.

Reply:
@briandowns I had to dig a little bit here. 

The `make forward-port` command calls a bash script in the [charts repo/scripts](https://github.com/rancher/charts/tree/dev-v2.9/scripts). It needs `yq` to operate. 

This feature has automated the `make forward-port` command. 

Can we create a separate issue for migrating the bash script `forward-script` inside the `charts-build-scripts`? 
Otherwise the scope of this feature will increase too much. 
And then I will remove the `os.Exec` commands for `yq` and do everything through go code.

Reply:
Let's do this work in another PR at some point.